README.md
[![Build Status](https://travis-ci.org/kianmeng/webservice-mocean.svg?branch=master)](https://travis-ci.org/kianmeng/webservice-mocean)
[![codecov](https://codecov.io/gh/kianmeng/webservice-mocean/branch/master/graph/badge.svg)](https://codecov.io/gh/kianmeng/webservice-mocean)
[![Coverage Status](https://coveralls.io/repos/kianmeng/webservice-mocean/badge.svg?branch=master)](https://coveralls.io/r/kianmeng/webservice-mocean?branch=master)
[![Kwalitee status](http://cpants.cpanauthors.org/dist/WebService-Mocean.png)](http://cpants.charsbar.org/dist/overview/WebService-Mocean)
[![Cpan license](https://img.shields.io/cpan/l/WebService-Mocean.svg)](https://metacpan.org/release/WebService-Mocean)
[![Cpan version](https://img.shields.io/cpan/v/WebService-Mocean.svg)](https://metacpan.org/release/WebService-Mocean)

# NAME

WebService::Mocean - Perl library for integration with MoceanSMS gateway,
https://moceanapi.com.

# SYNOPSIS

    use WebService::Mocean;
    my $mocean_api = WebService::Mocean->new(api_key => 'foo', api_secret => 'bar');

# DESCRIPTION

WebService::Mocean is Perl library for integration with MoceanSMS gateway,
https://moceanapi.com.

# DEVELOPMENT

Source repo at [https://github.com/kianmeng/webservice-mocean](https://github.com/kianmeng/webservice-mocean).

## Docker

If you have Docker installed, you can build your Docker container for this
project.

    $ docker build -t webservice-mocean .
    $ docker run -it -v $(pwd):/root webservice-mocean bash
    # cpanm --installdeps --notest .

## Milla

Setting up the required packages.

    $ milla authordeps --missing | cpanm
    $ milla listdeps --missing | cpanm

Check you code coverage.

    $ milla cover

Several ways to run the test.

    $ milla test
    $ milla test --author --release
    $ AUTHOR_TESTING=1 RELEASE_TESTING=1 milla test
    $ AUTHOR_TESTING=1 RELEASE_TESTING=1 milla run prove t/01_instantiation.t
    $ LOGGING=1 milla run prove t/t/02_request.t

Release the module.

    $ milla build
    $ milla release

# METHODS

## new($api\_key, $api\_secret, \[%$args\])

Construct a new WebService::Mocean instance. The api\_key and api\_secret is
compulsory fields. Optionally takes additional hash or hash reference.

    # Instantiate the class.
    my $mocean_api = WebService::Mocean->new(api_key => 'foo', api_secret => 'bar');

    # Alternative way.
    my $mocean_api = WebService::Mocean->new({api_key => 'foo', api_secret => 'bar'});

### api\_url

The URL of the API resource.

    # Instantiate the class by setting the URL of the API endpoints.
    my $mocean_api = WebService::Mocean->new({api_url => 'http://example.com/api/'});

    # Alternative way.
    my $mocean_api = WebService::Mocean->new(api_key => 'foo', api_secret => 'bar');
    $mocean_api->api_url('http://example.com/api/');

## sms->send($params)

Send Mobile Terminated (MT) message, which means the message is sent from
mobile SMS provider and terminated at the to the mobile phone.

    # Send sample SMS.
    my $response = $mocean_api->sms->send({
        'mocean-to' => '0123456789',
        'mocean-from' => 'ACME Ltd.',
        'mocean-text' => 'Hello'
    });

## sms->send\_verification\_code($params)

Send a random code for verification to a mobile number.

    my $response = $mocean_api->sms->send_verification_code({
        'mocean-to' => '0123456789',
        'mocean-brand' => 'ACME Ltd.',
    });

## sms->check\_verification\_code($params)

Check the verfication code received from your user.

    my $response = $mocean_api->sms->check_verification_code({
        'mocean-reqid' => '395935',
        'mocean-code' => '234839',
    });

## account->get\_balance()

Get your Mocean account balance.

    my $response = $mocean_api->account->get_balance();

## account->get\_pricing()

Get your Mocean account pricing and supported destination.

    my $response = $mocean_api->account->get_pricing();

## report->get\_message\_status($params)

Get the outbound SMS current status.

    my $response = $mocean_api->report->get_message_status({
        'mocean-msgid' => 123456
    });

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2018-2019 by Kian Meng, Ang.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)

# AUTHOR

Kian Meng, Ang <kianmeng@users.noreply.github.com>


Changes
Revision history for WebService-Mocean

{{$NEXT}}

0.04  2019-02-03 23:47:56 +08
        - Load modules dynamically
        - Skip test if authenticantion token is not set

0.03  2018-09-30 17:50:55 +08
        - Fix unused request params
        - Add query API web service calls

0.02  2018-09-30 00:51:14 +08
        - Check for required parameters
        - Add verification API web service calls

0.01  2018-08-26 18:08:27 +08
        - Initial release


lib/WebService/Mocean/Client.pm
package WebService::Mocean::Client;

use Moo;
use Types::Standard qw(Str Ref);
use Array::Utils qw(array_minus);

with 'Role::REST::Client';

our $VERSION = '0.05';

has api_url => (
    isa => Str,
    is => 'rw',
    default => sub { 'https://rest.moceanapi.com/rest/1' },
);

has api_key => (
    isa => Str,
    is => 'rw',
    required => 1
);

has api_secret => (
    isa => Str,
    is => 'rw',
    required => 1
);

has '_response_status' => (
    isa => Ref["HASH"],
    is => 'ro',
    init_arg => undef,
    default => sub {{
        0 => 'OK. No error encountered.',
        1 => 'Authorization failed. Invalid mocean-api-key or mocean-api-secret.',
        2 => 'Insufficient balance. Not enough credit in the account to send to at least one of the receivers.',
        4 => 'At least one of the destination numbers is not white listed.',
        5 => 'At least one of the destination numbers is black listed.',
        6 => 'No destination number specified.',
        8 => 'Sender ID not found.',
        9 => 'Invalid UDH field.',
        10 => 'Invalid mclass field.',
        17 => 'Invalid validity field.',
        19 => 'Invalid character set or message body.',
        20 => 'Insufficient headers for sending SMS.',
        23 => 'Empty mocean-text.',
        24 => 'Unknown error.',
        26 => 'Invalid schedule format. (Hint: must have leading zero for time.)',
        27 => 'Max number of receivers in a single request reached. Too many receivers in mocean-to field.',
        28 => 'Invalid destination number. Receiver is invalid after stripping all non-numerics.',
        29 => 'Message body is too long.',
        32 => 'Message throttled.',
        34 => 'Unknown request.',
        37 => 'Invalid sender length.',
        40 => 'System down for maintenance.',
        43 => 'SMS flooding detected.',
        44 => 'Invalid Sender ID.',
        45 => 'System error, please retry later.',
        48 => 'At least one of the senders is black listed.',
        49 => 'At least one of the senders is not white listed.',
        50 => 'Inappropriate content detected.',
    }}
);

has '_required_fields' => (
    isa => Ref["HASH"],
    is => 'ro',
    init_arg => undef,
    default => sub {{
        sms => [qw(mocean-from mocean-to mocean-text)],
        'verify/req' => [qw(mocean-to mocean-brand)],
        'verify/check' => [qw(mocean-reqid mocean-code)],
        'report/message' => [qw(mocean-msgid)],
        'account/balance' => [],
        'account/pricing' => [],
    }}
);

sub BUILD {
    my ($self, $args) = @_;

    $self->server($args->{api_url});
    $self->api_key($args->{api_key});
    $self->api_secret($args->{api_secret});

    $self->set_persistent_header(
        'User-Agent' => __PACKAGE__ . $WebService::Mocean::Client::VERSION
    );

    return $self;
}

sub request {
    my ($self, $command, $queries, $method) = @_;

    $command ||= q||;
    $queries ||= {};
    $method ||= 'get';

    $self->_check_required_params($command, $queries);

    my $params = $self->_auth_params();
    $queries = {%$queries, %$params};

    # In case the api_url was updated.
    $self->server($self->api_url);

    my $response_format = $queries->{'mocean-resp-format'} || 'xml';

    $self->type(qq|application/$response_format|);

    # Do not append '/' at the end of URL. Otherwise you will get HTTP 406
    # error.
    my $path = "/" . $command;

    my $response;
    if ($self->can($method)) {
        $response = $self->$method($path, $queries);
    }
    else {
        die "No such HTTP method: $method";
    }

    return $self->_response($response->data);
}

sub _response {
    my ($self, $response) = @_;

    my $response_code = $response->{'mocean-code'};
    my $response_status = $self->_response_status->{$response_code};

    $response->{'mocean-code-status'} = $response_status;

    return $response;
}

sub _auth_params {
    my ($self) = @_;

    return {
        'mocean-api-key' => $self->api_key,
        'mocean-api-secret' => $self->api_secret,
    };
}

sub _check_required_params {
    my ($self, $command, $params) = @_;

    my $required_fields = $self->_required_fields->{$command};

    die "Missing or invalid command : $command" if (!defined $required_fields);

    my @param_keys = keys %$params;
    my @missing = array_minus(@$required_fields, @param_keys);

    die "Missing required params: " . join(', ', @missing) if (scalar @missing);
}

1;


lib/WebService/Mocean.pm
package WebService::Mocean;

use namespace::clean;
use strictures 2;
use utf8;

use Module::Runtime qw(require_module);
use Moo;
use Types::Standard qw(Str InstanceOf);

use WebService::Mocean::Client;

our $VERSION = '0.05';

has api_key => (
    isa => Str,
    is => 'rw',
    required => 1
);

has api_secret => (
    isa => Str,
    is => 'rw',
    required => 1
);

has api_url => (
    isa => Str,
    is => 'rw',
    default => sub { 'https://rest.moceanapi.com/rest/1' },
);

has client => (
    is => 'lazy',
    isa => InstanceOf['WebService::Mocean::Client'],
);

sub _build_client {
    my $self = shift;

    my $client = WebService::Mocean::Client->new(
        api_key => $self->api_key,
        api_secret => $self->api_secret,
        api_url => $self->api_url,
    );

    return $client;
}

has sms => (
    is => 'lazy',
    default => sub { _delegate(shift, 'Sms') },
);

has account => (
    is => 'lazy',
    default => sub { _delegate(shift, 'Account') },
);

has report => (
    is => 'lazy',
    default => sub { _delegate(shift, 'Report') },
);

sub _delegate {
    my ($self, $module) = @_;

    my $package = __PACKAGE__ . "::$module";
    require_module $package;

    return $package->new(client => $self->client);
}

1;
__END__

=encoding utf-8

=head1 NAME

WebService::Mocean - Perl library for integration with MoceanSMS gateway,
https://moceanapi.com.

=head1 SYNOPSIS

  use WebService::Mocean;
  my $mocean_api = WebService::Mocean->new(api_key => 'foo', api_secret => 'bar');

=head1 DESCRIPTION

WebService::Mocean is Perl library for integration with MoceanSMS gateway,
https://moceanapi.com.

=head1 DEVELOPMENT

Source repo at L<https://github.com/kianmeng/webservice-mocean|https://github.com/kianmeng/webservice-mocean>.

=head2 Docker

If you have Docker installed, you can build your Docker container for this
project.

    $ docker build -t webservice-mocean .
    $ docker run -it -v $(pwd):/root webservice-mocean bash
    # cpanm --installdeps --notest .

=head2 Milla

Setting up the required packages.

    $ milla authordeps --missing | cpanm
    $ milla listdeps --missing | cpanm

Check you code coverage.

    $ milla cover

Several ways to run the test.

    $ milla test
    $ milla test --author --release
    $ AUTHOR_TESTING=1 RELEASE_TESTING=1 milla test
    $ AUTHOR_TESTING=1 RELEASE_TESTING=1 milla run prove t/01_instantiation.t
    $ LOGGING=1 milla run prove t/t/02_request.t

Release the module.

    $ milla build
    $ milla release

=head1 METHODS

=head2 new($api_key, $api_secret, [%$args])

Construct a new WebService::Mocean instance. The api_key and api_secret is
compulsory fields. Optionally takes additional hash or hash reference.

    # Instantiate the class.
    my $mocean_api = WebService::Mocean->new(api_key => 'foo', api_secret => 'bar');

    # Alternative way.
    my $mocean_api = WebService::Mocean->new({api_key => 'foo', api_secret => 'bar'});

=head3 api_url

The URL of the API resource.

    # Instantiate the class by setting the URL of the API endpoints.
    my $mocean_api = WebService::Mocean->new({api_url => 'http://example.com/api/'});

    # Alternative way.
    my $mocean_api = WebService::Mocean->new(api_key => 'foo', api_secret => 'bar');
    $mocean_api->api_url('http://example.com/api/');

=head2 sms->send($params)

Send Mobile Terminated (MT) message, which means the message is sent from
mobile SMS provider and terminated at the to the mobile phone.

    # Send sample SMS.
    my $response = $mocean_api->sms->send({
        'mocean-to' => '0123456789',
        'mocean-from' => 'ACME Ltd.',
        'mocean-text' => 'Hello'
    });

=head2 sms->send_verification_code($params)

Send a random code for verification to a mobile number.

    my $response = $mocean_api->sms->send_verification_code({
        'mocean-to' => '0123456789',
        'mocean-brand' => 'ACME Ltd.',
    });

=head2 sms->check_verification_code($params)

Check the verfication code received from your user.

    my $response = $mocean_api->sms->check_verification_code({
        'mocean-reqid' => '395935',
        'mocean-code' => '234839',
    });

=head2 account->get_balance()

Get your Mocean account balance.

    my $response = $mocean_api->account->get_balance();

=head2 account->get_pricing()

Get your Mocean account pricing and supported destination.

    my $response = $mocean_api->account->get_pricing();

=head2 report->get_message_status($params)

Get the outbound SMS current status.

    my $response = $mocean_api->report->get_message_status({
        'mocean-msgid' => 123456
    });

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2018 by Kian Meng, Ang.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)

=head1 AUTHOR

Kian Meng, Ang E<lt>kianmeng@users.noreply.github.comE<gt>
